File: packages/kits/GeniusLocusKit/Tests/GeniusLocusKitTests/SharedContentMigrationTests.swift
// SharedContentMigrationTests.swift
//
// Resumable legacy-migration coverage (GLK shared-content 1.1, P4).
//
// Builds a REAL legacy estate shape in place — captured Drawers plus the
// pre-cutover copy lane (chunks + corpus_metadata + chunk-keyed vectors,
// with the legacy "CorpusKit" schema version recorded) — then proves:
// selective deletion (protected rows survive byte-identically), declared
// schema retirement, streamed Drawer-keyed rebuild, verification, orphan
// fail-dark, fault-injected resume equivalence, and the dark-lane gate.

import Testing
import Foundation
import CorpusKit
import CorpusKitProviders
import LocusKit
import PersistenceKit
import PersistenceKitSQLite
import VectorKit
import EngramLib
import SubstrateTypes

@testable import GeniusLocusKit

@Suite("SharedContentMigration", .serialized)
struct SharedContentMigrationTests {

    private let now = Date(timeIntervalSince1970: 1_700_000_000)

    private func scratchStorage() throws -> (any Storage, URL) {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("glk-scm-\(UUID().uuidString).sqlite3")
        let storage = try SQLiteStorage(configuration: EstateConfiguration(
            estateID: UUID(), backend: .sqlite(url: url, busyTimeout: 5.0)))
        return (storage, url)
    }

    /// Open a GLK estate (no engine registered — kit.open wires none),
    /// capture drawers, and overlay the LEGACY copy lane exactly as the
    /// pre-cutover engine produced it.
    private func makeLegacyEstate(
        drawerContents: [String],
        includeOrphanChunk: Bool = false,
        seedProtectedVector: Bool = true
    ) async throws -> (kit: GeniusLocusKit, handle: EstateHandle,
                       storage: any Storage, drawerIDs: [String], url: URL) {
        let kit = GeniusLocusKit()
        let owner = OwnerCredentials(ownerIdentifier: "scm-owner")
        let (storage, url) = try scratchStorage()
        _ = try await LocusKit.Estate.create(storage: storage, owner: owner)
        let handle = try await kit.open(storage: storage, owner: owner)

        var drawerIDs: [String] = []
        for content in drawerContents {
            let frame = CaptureFrame(
                content: content, channel: .typed, room: "scm",
                latticeAnchor: LatticeAnchor(udcCode: "004"),
                addedBy: "scm-test", embeddingModelID: "scm-v1")
            let drawer = try await kit.capture(handle, frame)
            drawerIDs.append(drawer.id)
        }

        // Overlay the legacy copy lane: BundleStore schema (registers the
        // legacy "CorpusKit" v3), one chunk row per drawer with VERBATIM
        // text, and one chunk-keyed vector row per chunk.
        try await storage.migrate(to: BundleStore.schemaDeclaration)
        try await storage.migrate(to: RemovedSourceStore.schemaDeclaration)
        try await storage.migrate(to: VectorStore.schemaDeclaration)
        for (index, content) in drawerContents.enumerated() {
            let chunkID = Chunk.deriveID(
                sourceID: drawerIDs[index], startOffset: 0, text: content)
            _ = try await storage.rowStore.insert(table: "chunks", values: [
                "id": .uuid(chunkID),
                "source_id": .text(drawerIDs[index]),
                "start_offset": .int(0),
                "length": .int(Int64(content.count)),
                "text": .text(content),
                "hlc": .hlc(HLC(physicalTime: Int64(index + 1), logicalCount: 0, nodeID: 1)),
                "metadata": .json(Data("{}".utf8)),
                "created_at": .timestamp(now),
                "ext": .null
            ])
            _ = try await storage.rowStore.insert(table: "vectors", values: [
                "id": .uuid(UUID()),
                "item_id": .text(chunkID.uuidString),
                "vector_index": .int(0),
                "model_id": .text("corpus-deterministic-v1"),
                "model_version": .text("1.0.0"),
                "kind": .int(0),
                "dim": .int(256),
                "payload": .blob(Data(repeating: UInt8(index + 1), count: 32)),
                "scale": .null,
                "filed_at": .timestamp(now)
            ])
        }
        if includeOrphanChunk {
            let orphanID = Chunk.deriveID(
                sourceID: "ghost-drawer", startOffset: 0, text: "orphaned text")
            _ = try await storage.rowStore.insert(table: "chunks", values: [
                "id": .uuid(orphanID),
                "source_id": .text("ghost-drawer"),
                "start_offset": .int(0),
                "length": .int(13),
                "text": .text("orphaned text"),
                "hlc": .hlc(HLC(physicalTime: 99, logicalCount: 0, nodeID: 1)),
                "metadata": .json(Data("{}".utf8)),
                "created_at": .timestamp(now),
                "ext": .null
            ])
        }
        if seedProtectedVector {
            // An UNRELATED lane's representation, keyed by a Drawer ID under
            // a model the corpus never claims — must survive byte-identically.
            _ = try await storage.rowStore.insert(table: "vectors", values: [
                "id": .uuid(UUID()),
                "item_id": .text(drawerIDs[0]),
                "vector_index": .int(0),
                "model_id": .text("unrelated-lane-v1"),
                "model_version": .text("1.0.0"),
                "kind": .int(0),
                "dim": .int(256),
                "payload": .blob(Data(repeating: 0xEE, count: 32)),
                "scale": .null,
                "filed_at": .timestamp(now)
            ])
        }
        return (kit, handle, storage, drawerIDs, url)
    }

    private func protectedVectorPayload(_ storage: any Storage, itemID: String) async throws -> Data? {
        let rows = try await storage.rowStore.query(
            table: "vectors",
            where: .and([
                .eq(Column(table: "vectors", name: "item_id"), .text(itemID)),
                .eq(Column(table: "vectors", name: "model_id"), .text("unrelated-lane-v1"))
            ]),
            orderBy: [], limit: nil, offset: nil)
        guard let row = rows.first, case let .blob(data)? = row["payload"] else { return nil }
        return data
    }

    // MARK: - Fresh estates bypass

    @Test func freshEstateBypassesAndNeverCreatesLegacyTables() async throws {
        let kit = GeniusLocusKit()
        let owner = OwnerCredentials(ownerIdentifier: "scm-fresh")
        let (storage, url) = try scratchStorage()
        defer { try? FileManager.default.removeItem(at: url) }
        _ = try await LocusKit.Estate.create(storage: storage, owner: owner)
        let handle = try await kit.open(storage: storage, owner: owner)

        let report = try await kit.runSharedContentMigration(handle: handle, now: now)
        #expect(report.state == .complete)
        #expect(report.legacyChunkCount == 0)
        // Bypass creates no legacy tables.
        await #expect(throws: (any Error).self) {
            _ = try await storage.rowStore.count(table: "chunks", where: nil)
        }
    }

    // MARK: - Happy path

    @Test func legacyEstateMigratesSelectivelyAndVerifies() async throws {
        let contents = [
            "The migration must keep this drawer intact.",
            "A second drawer with distinct content for recall.",
            "Third drawer about physical page reclamation."
        ]
        let (kit, handle, storage, drawerIDs, url) =
            try await makeLegacyEstate(drawerContents: contents)
        defer { try? FileManager.default.removeItem(at: url) }

        let protectedBefore = try await protectedVectorPayload(storage, itemID: drawerIDs[0])
        #expect(protectedBefore != nil)

        let report = try await kit.runSharedContentMigration(handle: handle, now: now)
        #expect(report.state == .reclaimPending)
        #expect(report.legacyChunkCount == 3)
        #expect(report.legacyVectorKeyCount == 3)
        #expect(report.rebuiltContentCount == 3)

        // Copy-lane tables are GONE (declared dropTable retirement).
        await #expect(throws: (any Error).self) {
            _ = try await storage.rowStore.count(table: "chunks", where: nil)
        }
        await #expect(throws: (any Error).self) {
            _ = try await storage.rowStore.count(table: "corpus_metadata", where: nil)
        }

        // No chunk-keyed vector survives; Drawer-keyed rows exist for every
        // drawer; the unrelated lane's row survived BYTE-IDENTICALLY.
        let vectorRows = try await storage.rowStore.query(
            table: "vectors", where: nil, orderBy: [], limit: nil, offset: nil)
        var itemIDs: Set<String> = []
        for row in vectorRows {
            if case let .text(item)? = row["item_id"] { itemIDs.insert(item) }
        }
        for id in drawerIDs { #expect(itemIDs.contains(id)) }
        #expect(try await protectedVectorPayload(storage, itemID: drawerIDs[0])
            == protectedBefore)

        // The rebuilt lane hydrates directly: BM25 hits ARE drawer IDs.
        let estateObj = try await kit.estate(for: handle)
        let engine = try await CorpusContentEngine(
            storage: storage,
            configuration: CorpusContentConfiguration(
                mode: .attached, indexUnit: .wholeContent),
            source: LocusDrawerCorpusContentSource(estate: estateObj))
        let hits = try await engine.bm25TopK(query: "page reclamation", limit: 5)
        #expect(hits.first?.id == drawerIDs[2])

        // Idempotent re-run: same terminal state, nothing re-deleted.
        let rerun = try await kit.runSharedContentMigration(handle: handle, now: now)
        #expect(rerun.state == .reclaimPending)

        // Reclaim completion runs the PHYSICAL reclamation (WAL checkpoint
        // + VACUUM): the retired legacy tables' pages must actually leave
        // the filesystem, and the status surface must report the outcome.
        let statusBefore = try await kit.sharedContentReclaimStatus(handle: handle)
        #expect(statusBefore.state == .reclaimPending)
        let maintenance = try #require(
            try await kit.completeSharedContentReclaim(handle: handle, now: now))
        #expect(maintenance.performed)
        #expect(maintenance.backend == "sqlite")
        #expect(maintenance.freelistPagesAfter == 0)
        #expect(maintenance.reclaimedBytes > 0)
        #expect(maintenance.fileSizeBytesAfter + maintenance.walBytesAfter
            < maintenance.fileSizeBytesBefore + maintenance.walBytesBefore)
        #expect(try await kit.sharedContentMigrationState(handle: handle) == .complete)
        let statusAfter = try await kit.sharedContentReclaimStatus(handle: handle)
        #expect(statusAfter.state == .complete)
        #expect(statusAfter.reclaimedBytes == maintenance.reclaimedBytes)

        // Recall still works over the vacuumed file: the reclamation freed
        // pages, never derived state.
        let postHits = try await engine.bm25TopK(query: "page reclamation", limit: 5)
        #expect(postHits.first?.id == drawerIDs[2])
    }

    // MARK: - Ensemble upgrade (add a provider to a migrated estate)

    @Test func ensembleUpgradeCoversAddedProviderAndStaysDarkUntilThen() async throws {
        let contents = [
            "Upgrade fixture drawer one about estates.",
            "Upgrade fixture drawer two about coverage."
        ]
        let (kit, handle, storage, drawerIDs, url) =
            try await makeLegacyEstate(drawerContents: contents)
        defer { try? FileManager.default.removeItem(at: url) }

        // Migrate under a ONE-provider configuration and finish reclaim.
        let small: [EmbeddingModel] = [.deterministic]
        let first = try await kit.runSharedContentMigration(
            handle: handle, now: now, embeddingModels: small)
        #expect(first.state == .reclaimPending)
        _ = try await kit.completeSharedContentReclaim(handle: handle, now: now)
        let smallFP = CorpusContentEngine.configurationFingerprint(
            mode: .attached, models: small)
        #expect(!(await kit.sharedContentLaneMustStayDark(
            storage: storage, wiredFingerprint: smallFP)))

        // Wire a LARGER ensemble: the completed record is NOT trusted —
        // the lane goes dark and the estate enters a follow-on upgrade.
        let big: [EmbeddingModel] = [
            .deterministic,
            .randomIndexing(provider: RandomIndexingProvider())
        ]
        let bigFP = CorpusContentEngine.configurationFingerprint(
            mode: .attached, models: big)
        #expect(await kit.sharedContentLaneMustStayDark(
            storage: storage, wiredFingerprint: bigFP))

        // The upgrade trains the ADDED provider, backfills ONLY its missing
        // coverage, re-verifies, and restamps the fingerprint.
        let upgraded = try await kit.runSharedContentMigration(
            handle: handle, now: now, embeddingModels: big)
        #expect(upgraded.state == .reclaimPending)
        _ = try await kit.completeSharedContentReclaim(handle: handle, now: now)
        #expect(!(await kit.sharedContentLaneMustStayDark(
            storage: storage, wiredFingerprint: bigFP)))

        // Every provider covers every drawer, and BM25 still serves.
        let estateObj = try await kit.estate(for: handle)
        let engine = try await CorpusContentEngine(
            storage: storage,
            configuration: CorpusContentConfiguration(
                mode: .attached, indexUnit: .wholeContent),
            source: LocusDrawerCorpusContentSource(estate: estateObj),
            models: big)
        #expect(try await engine.coveredCount(modelID: "corpus-deterministic-v1")
            == drawerIDs.count)
        #expect(try await engine.coveredCount(modelID: "random-indexing-v1")
            == drawerIDs.count)
        let hits = try await engine.bm25TopK(query: "coverage", limit: 5)
        #expect(hits.first?.id == drawerIDs[1])
    }

    // MARK: - Orphan fail-dark

    @Test func orphanedLegacySourceStopsDarkBeforeAnyDeletion() async throws {
        let (kit, handle, storage, _, url) = try await makeLegacyEstate(
            drawerContents: ["Content with a valid drawer."],
            includeOrphanChunk: true)
        defer { try? FileManager.default.removeItem(at: url) }

        await #expect(throws: SharedContentMigrationError.self) {
            try await kit.runSharedContentMigration(handle: handle, now: now)
        }
        // NOTHING was dropped — the copy lane is intact for diagnosis.
        #expect(try await storage.rowStore.count(table: "chunks", where: nil) == 2)
        // And the lane stays dark.
        #expect(await kit.sharedContentLaneMustStayDark(storage: storage))
    }

    // MARK: - Fault-injected resume equivalence

    @Test func faultAfterEveryStateResumesToTheSameOutcome() async throws {
        let contents = ["Resume drawer one.", "Resume drawer two."]

        // Uninterrupted reference run.
        let (kitRef, handleRef, storageRef, _, urlRef) =
            try await makeLegacyEstate(drawerContents: contents)
        defer { try? FileManager.default.removeItem(at: urlRef) }
        let reference = try await kitRef.runSharedContentMigration(handle: handleRef, now: now)
        let referenceInventory = try await DatabaseInventory.capture(
            storage: storageRef, tables: ["vectors", "corpus_index_state"],
            excludingColumns: ["vectors": ["id"]])

        let faultStates: [SharedContentMigrationState] = [
            .discovered, .canonicalValidated, .legacyInventoryCaptured,
            .legacyDerivedCleared, .legacySchemaRetired, .drawerIndexRebuilt,
            .verified
        ]
        for fault in faultStates {
            let (kit, handle, storage, _, url) =
                try await makeLegacyEstate(drawerContents: contents)
            defer { try? FileManager.default.removeItem(at: url) }

            await kit._setSharedContentFault(fault)
            await #expect(throws: SharedContentMigrationError.self) {
                try await kit.runSharedContentMigration(handle: handle, now: now)
            }
            // Resume: the rerun continues from the persisted state and
            // reaches the SAME outcome as the uninterrupted run.
            let resumed = try await kit.runSharedContentMigration(handle: handle, now: now)
            #expect(resumed.state == reference.state,
                    "fault after \(fault.rawValue) must resume to the reference state")
            #expect(resumed.legacyChunkCount == reference.legacyChunkCount)
            #expect(resumed.rebuiltContentCount == reference.rebuiltContentCount)
            let inventory = try await DatabaseInventory.capture(
                storage: storage, tables: ["vectors", "corpus_index_state"],
                excludingColumns: ["vectors": ["id"]])
            // Logical derived bytes match the uninterrupted run (surrogate
            // vector ids and drawer UUIDs differ per estate, so compare
            // counts, not folds, across DIFFERENT estates; folds prove
            // intra-estate stability elsewhere).
            #expect(inventory.map(\.rowCount) == referenceInventory.map(\.rowCount),
                    "fault after \(fault.rawValue): row populations must match the uninterrupted run")
        }
    }

    // MARK: - Dark-lane gate

    @Test func legacyEstateKeepsCorpusLaneDarkUntilMigrated() async throws {
        let (kit, handle, storage, _, url) = try await makeLegacyEstate(
            drawerContents: ["Dark lane drawer."])
        defer { try? FileManager.default.removeItem(at: url) }
        #expect(await kit.sharedContentLaneMustStayDark(storage: storage))
        _ = try await kit.runSharedContentMigration(handle: handle, now: now)
        #expect(!(await kit.sharedContentLaneMustStayDark(storage: storage)))
        _ = handle
    }
}

File: packages/kits/GeniusLocusKit/rust/tests/shared_content_migration_tests.rs
//! Resumable legacy-migration coverage (GLK shared-content 1.1, P4).
//! Rust twin of the Swift `SharedContentMigrationTests`.

use std::collections::BTreeMap;
use std::sync::Arc;

use corpus_kit::{Chunk, CorpusContentConfiguration, CorpusContentEngine,
    CorpusIndexUnitPolicy, CorpusOperatingMode, EmbeddingModelConfig};
use genius_locus_kit::intake::LocusDrawerContentSource;
use genius_locus_kit::shared_content_migration::{
    SharedContentMigrationError, SharedContentMigrationState,
};
use genius_locus_kit::EstateCoordinator;
use locus_kit::drawer_operational::CaptureChannel;
use locus_kit::drawer_store::DrawerStore;
use locus_kit::drawer_store_inmemory::InMemoryDrawerStore;
use locus_kit::drawer_store_sqlite::SqliteDrawerStore;
use locus_kit::estate_types::{LatticeAnchor, OwnerCredentials};
use locus_kit::frames::CaptureFrame;
use persistence_kit::database_inventory::capture_inventory;
use persistence_kit::inmemory::InMemoryStorage;
use persistence_kit::{Storage, TypedValue};
use substrate_types::hlc::HLC;

const NOW: i64 = 1_700_000_000_000;

fn cap_frame(content: &str) -> CaptureFrame {
    CaptureFrame::new(
        content,
        CaptureChannel::Typed,
        "scm",
        LatticeAnchor::udc("004"),
        "scm-test",
        "scm-v1",
    )
}

struct LegacyEstate {
    coord: EstateCoordinator,
    handle: genius_locus_kit::handle::EstateHandle,
    storage: Arc<dyn Storage>,
    drawer_ids: Vec<String>,
}

fn make_legacy_estate(contents: &[&str], include_orphan: bool, seed_protected: bool) -> LegacyEstate {
    // SQLite-backed like the Swift twin: the Rust BM25 sidecar persists in
    // the estate FILE, so a verification engine built after the migration
    // sees the rebuilt postings (an InMemory estate's sidecar is
    // per-connection by design).
    let mut coord = EstateCoordinator::new();
    let path = std::env::temp_dir()
        .join(format!("glk-scm-{}.sqlite", uuid::Uuid::new_v4()));
    let sqlite_store = SqliteDrawerStore::from_path(
        path.to_string_lossy().as_ref(), NOW, None, 5.0)
        .expect("SqliteDrawerStore::from_path");
    let store: Arc<dyn DrawerStore> = Arc::new(sqlite_store);
    let storage: Arc<dyn Storage> = store.storage().expect("sqlite store exposes storage");
    let handle = coord
        .open(store, OwnerCredentials::new("scm-owner"), 0, 100)
        .expect("open");

    let mut drawer_ids = Vec::new();
    for content in contents {
        let drawer = coord.capture(&handle, cap_frame(content), NOW).expect("capture");
        drawer_ids.push(drawer.id);
    }

    // Overlay the legacy copy lane.
    storage
        .migrate(&corpus_kit::BundleStore::schema_declaration())
        .expect("legacy chunks schema");
    storage
        .migrate(&corpus_kit::removed_source_store::RemovedSourceStore::schema_declaration())
        .expect("legacy removed_sources schema");
    storage
        .migrate(&vectorkit::VectorStore::schema_declaration())
        .expect("vectors schema");
    let row_store = storage.row_store();
    for (index, content) in contents.iter().enumerate() {
        let chunk_id = Chunk::derive_id(&drawer_ids[index], 0, content);
        let mut chunk: BTreeMap<String, TypedValue> = BTreeMap::new();
        chunk.insert("id".into(), TypedValue::Uuid(chunk_id));
        chunk.insert("source_id".into(), TypedValue::Text(drawer_ids[index].clone()));
        chunk.insert("start_offset".into(), TypedValue::Int(0));
        chunk.insert("length".into(), TypedValue::Int(content.len() as i64));
        chunk.insert("text".into(), TypedValue::Text(content.to_string()));
        chunk.insert("hlc".into(), TypedValue::Hlc(HLC {
            physical_time: (index + 1) as i64, logical_count: 0, node_id: 1 }));
        chunk.insert("metadata".into(), TypedValue::Json(b"{}".to_vec()));
        chunk.insert("created_at".into(), TypedValue::Timestamp(NOW));
        chunk.insert("ext".into(), TypedValue::Null);
        row_store.insert("chunks", chunk).expect("insert chunk");

        let mut vector: BTreeMap<String, TypedValue> = BTreeMap::new();
        vector.insert("id".into(), TypedValue::Uuid(uuid::Uuid::new_v4()));
        vector.insert("item_id".into(),
            TypedValue::Text(chunk_id.to_string().to_uppercase()));
        vector.insert("vector_index".into(), TypedValue::Int(0));
        vector.insert("model_id".into(), TypedValue::Text("corpus-deterministic-v1".into()));
        vector.insert("model_version".into(), TypedValue::Text("1.0.0".into()));
        vector.insert("kind".into(), TypedValue::Int(0));
        vector.insert("dim".into(), TypedValue::Int(256));
        vector.insert("payload".into(), TypedValue::Blob(vec![(index + 1) as u8; 32]));
        vector.insert("scale".into(), TypedValue::Null);
        vector.insert("filed_at".into(), TypedValue::Timestamp(NOW));
        row_store.insert("vectors", vector).expect("insert vector");
    }
    if include_orphan {
        let orphan_id = Chunk::derive_id("ghost-drawer", 0, "orphaned text");
        let mut chunk: BTreeMap<String, TypedValue> = BTreeMap::new();
        chunk.insert("id".into(), TypedValue::Uuid(orphan_id));
        chunk.insert("source_id".into(), TypedValue::Text("ghost-drawer".into()));
        chunk.insert("start_offset".into(), TypedValue::Int(0));
        chunk.insert("length".into(), TypedValue::Int(13));
        chunk.insert("text".into(), TypedValue::Text("orphaned text".into()));
        chunk.insert("hlc".into(), TypedValue::Hlc(HLC {
            physical_time: 99, logical_count: 0, node_id: 1 }));
        chunk.insert("metadata".into(), TypedValue::Json(b"{}".to_vec()));
        chunk.insert("created_at".into(), TypedValue::Timestamp(NOW));
        chunk.insert("ext".into(), TypedValue::Null);
        row_store.insert("chunks", chunk).expect("insert orphan");
    }
    if seed_protected {
        let mut vector: BTreeMap<String, TypedValue> = BTreeMap::new();
        vector.insert("id".into(), TypedValue::Uuid(uuid::Uuid::new_v4()));
        vector.insert("item_id".into(), TypedValue::Text(drawer_ids[0].clone()));
        vector.insert("vector_index".into(), TypedValue::Int(0));
        vector.insert("model_id".into(), TypedValue::Text("unrelated-lane-v1".into()));
        vector.insert("model_version".into(), TypedValue::Text("1.0.0".into()));
        vector.insert("kind".into(), TypedValue::Int(0));
        vector.insert("dim".into(), TypedValue::Int(256));
        vector.insert("payload".into(), TypedValue::Blob(vec![0xEE; 32]));
        vector.insert("scale".into(), TypedValue::Null);
        vector.insert("filed_at".into(), TypedValue::Timestamp(NOW));
        row_store.insert("vectors", vector).expect("insert protected");
    }
    LegacyEstate { coord, handle, storage, drawer_ids }
}

fn protected_payload(storage: &Arc<dyn Storage>, item_id: &str) -> Option<Vec<u8>> {
    let rows = storage
        .row_store()
        .query("vectors", None, &[], None, None)
        .ok()?;
    for row in rows {
        if let (Some(TypedValue::Text(item)), Some(TypedValue::Text(model)), Some(TypedValue::Blob(payload))) =
            (row.get("item_id"), row.get("model_id"), row.get("payload"))
        {
            if item == item_id && model == "unrelated-lane-v1" {
                return Some(payload.clone());
            }
        }
    }
    None
}

#[test]
fn fresh_estate_bypasses_and_never_creates_legacy_tables() {
    let mut coord = EstateCoordinator::new();
    let backing = Arc::new(InMemoryStorage::with_estate(uuid::Uuid::new_v4()));
    let store: Arc<dyn DrawerStore> = Arc::new(
        InMemoryDrawerStore::with_storage(Arc::clone(&backing), NOW, None).unwrap(),
    );
    let handle = coord
        .open(store, OwnerCredentials::new("scm-fresh"), 0, 100)
        .expect("open");
    let report = coord
        .run_shared_content_migration(&handle, NOW)
        .expect("fresh migration");
    assert_eq!(report.state, SharedContentMigrationState::Complete);
    assert_eq!(report.legacy_chunk_count, 0);
    let storage: Arc<dyn Storage> = backing;
    assert!(storage.row_store().count("chunks", None).is_err());
}

#[test]
fn legacy_estate_migrates_selectively_and_verifies() {
    let contents = [
        "The migration must keep this drawer intact.",
        "A second drawer with distinct content for recall.",
        "Third drawer about physical page reclamation.",
    ];
    let mut est = make_legacy_estate(&contents, false, true);
    let protected_before =
        protected_payload(&est.storage, &est.drawer_ids[0]).expect("protected row");

    let report = est
        .coord
        .run_shared_content_migration(&est.handle, NOW)
        .expect("migration");
    assert_eq!(report.state, SharedContentMigrationState::ReclaimPending);
    assert_eq!(report.legacy_chunk_count, 3);
    assert_eq!(report.legacy_vector_key_count, 3);
    assert_eq!(report.rebuilt_content_count, 3);

    assert!(est.storage.row_store().count("chunks", None).is_err());
    assert!(est.storage.row_store().count("corpus_metadata", None).is_err());

    // Drawer-keyed rows exist; protected row survived byte-identically.
    assert_eq!(
        protected_payload(&est.storage, &est.drawer_ids[0]),
        Some(protected_before)
    );

    // Rebuilt lane hydrates directly: BM25 hits ARE drawer IDs.
    let estate = est.coord.estate_for(&est.handle).unwrap().clone();
    let engine = CorpusContentEngine::open(
        Arc::clone(&est.storage),
        CorpusContentConfiguration::new(
            CorpusOperatingMode::Attached,
            CorpusIndexUnitPolicy::WholeContent,
        )
        .unwrap(),
        Arc::new(LocusDrawerContentSource::new(estate)),
        vec![EmbeddingModelConfig::Deterministic],
    )
    .expect("engine");
    let hits = engine.bm25_top_k("page reclamation", 5).expect("bm25");
    assert_eq!(hits.first().map(|(id, _)| id.as_str()),
               Some(est.drawer_ids[2].as_str()));

    // Idempotent re-run, then reclaim completion.
    let rerun = est
        .coord
        .run_shared_content_migration(&est.handle, NOW)
        .expect("re-run");
    assert_eq!(rerun.state, SharedContentMigrationState::ReclaimPending);
    let status_before = est.coord.shared_content_reclaim_status(&est.handle);
    assert_eq!(
        status_before.state,
        Some(SharedContentMigrationState::ReclaimPending)
    );
    // Reclaim completion runs the PHYSICAL reclamation (WAL checkpoint +
    // VACUUM): the retired legacy tables' pages must actually leave the
    // filesystem, and the status surface must report the outcome.
    let maintenance = est
        .coord
        .complete_shared_content_reclaim(&est.handle, NOW)
        .expect("reclaim complete")
        .expect("maintenance report");
    assert!(maintenance.performed);
    assert_eq!(maintenance.backend, "sqlite");
    assert_eq!(maintenance.freelist_pages_after, 0);
    assert!(maintenance.reclaimed_bytes > 0);
    assert!(
        maintenance.file_size_bytes_after + maintenance.wal_bytes_after
            < maintenance.file_size_bytes_before + maintenance.wal_bytes_before
    );
    assert_eq!(
        est.coord.shared_content_migration_state(&est.handle),
        Some(SharedContentMigrationState::Complete)
    );
    let status_after = est.coord.shared_content_reclaim_status(&est.handle);
    assert_eq!(status_after.state, Some(SharedContentMigrationState::Complete));
    assert_eq!(status_after.reclaimed_bytes, Some(maintenance.reclaimed_bytes));

    // Recall still works over the vacuumed file: the reclamation freed
    // pages, never derived state.
    let post_hits = engine.bm25_top_k("page reclamation", 5).expect("bm25");
    assert_eq!(post_hits.first().map(|(id, _)| id.as_str()),
               Some(est.drawer_ids[2].as_str()));
}

#[test]
fn orphaned_legacy_source_stops_dark_before_any_deletion() {
    let mut est = make_legacy_estate(&["Content with a valid drawer."], true, true);
    let result = est.coord.run_shared_content_migration(&est.handle, NOW);
    assert!(matches!(
        result,
        Err(SharedContentMigrationError::OrphanedLegacySources { .. })
    ));
    assert_eq!(est.storage.row_store().count("chunks", None).unwrap(), 2);
    assert!(EstateCoordinator::shared_content_lane_must_stay_dark(&est.storage));
}

#[test]
fn fault_after_every_state_resumes_to_the_same_outcome() {
    let contents = ["Resume drawer one.", "Resume drawer two."];

    let mut reference = make_legacy_estate(&contents, false, true);
    let reference_report = reference
        .coord
        .run_shared_content_migration(&reference.handle, NOW)
        .expect("reference run");
    let reference_inventory = capture_inventory(
        &reference.storage,
        &["vectors", "corpus_index_state"],
        &BTreeMap::new(),
    )
    .expect("reference inventory");

    let fault_states = [
        SharedContentMigrationState::Discovered,
        SharedContentMigrationState::CanonicalValidated,
        SharedContentMigrationState::LegacyInventoryCaptured,
        SharedContentMigrationState::LegacyDerivedCleared,
        SharedContentMigrationState::LegacySchemaRetired,
        SharedContentMigrationState::DrawerIndexRebuilt,
        SharedContentMigrationState::Verified,
    ];
    for fault in fault_states {
        let mut est = make_legacy_estate(&contents, false, true);
        est.coord.set_shared_content_fault(Some(fault));
        let interrupted = est.coord.run_shared_content_migration(&est.handle, NOW);
        assert!(
            matches!(
                interrupted,
                Err(SharedContentMigrationError::InjectedFault { .. })
            ),
            "fault after {fault:?} must interrupt"
        );
        let resumed = est
            .coord
            .run_shared_content_migration(&est.handle, NOW)
            .expect("resume");
        assert_eq!(resumed.state, reference_report.state,
                   "fault after {fault:?} must resume to the reference state");
        assert_eq!(resumed.legacy_chunk_count, reference_report.legacy_chunk_count);
        assert_eq!(resumed.rebuilt_content_count, reference_report.rebuilt_content_count);
        let inventory = capture_inventory(
            &est.storage,
            &["vectors", "corpus_index_state"],
            &BTreeMap::new(),
        )
        .expect("inventory");
        let counts: Vec<usize> = inventory.iter().map(|i| i.row_count).collect();
        let reference_counts: Vec<usize> =
            reference_inventory.iter().map(|i| i.row_count).collect();
        assert_eq!(counts, reference_counts,
                   "fault after {fault:?}: row populations must match the uninterrupted run");
    }
}

#[test]
fn legacy_estate_keeps_corpus_lane_dark_until_migrated() {
    let mut est = make_legacy_estate(&["Dark lane drawer."], false, false);
    assert!(EstateCoordinator::shared_content_lane_must_stay_dark(&est.storage));
    est.coord
        .run_shared_content_migration(&est.handle, NOW)
        .expect("migration");
    assert!(!EstateCoordinator::shared_content_lane_must_stay_dark(&est.storage));
}
